blog/2024-03-27-reductstore-vs-timescale/index.md

```diff
@@ -0,0 +1,220 @@
+
+<!--truncate-->
+
+
+## **Read/Write Blob Data With** TimescaleDB
+
+Since TimescaleDB is a part of PostgreSQL, you can use the **[psycopg](https://pypi.org/project/psycopg2/)** adapter for Python to manage the database and stream the data. In this section, we initialize the TimescaleDB extension, recreate the benchmark database, and create a table for our blob data. Once the table is prepared, we write a chunk of binary data with the current time `BLOB_COUNT` times.
+
+```python
+import psycopg2
+from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
+
+BLOB_SIZE = 100_000
+CHUNK = random.randbytes(BLOB_SIZE)
+
+HOST = "localhost"
+CONNECTION = f"postgresql://postgres:postgres@{HOST}:5432"
+
+def setup_database():
+    con = psycopg2.connect(CONNECTION)
+    con.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+    cur = con.cursor()
+    cur.execute("CREATE EXTENSION IF NOT EXISTS timescaledb CASCADE;")
+    cur.execute(f"DROP DATABASE IF EXISTS benchmark")
+    cur.execute(f"CREATE DATABASE benchmark")
+    con.commit()
+    con.close()
+
+def write_to_timescale():
+    setup_database()
+
+    with psycopg2.connect(CONNECTION + "/benchmark") as con:
+        with con.cursor() as cur:
+            cur.execute(
+                f"""CREATE TABLE data (
+                       time TIMESTAMPTZ NOT NULL,
+                       blob_data BYTEA NOT NULL);
+                """
+            )
+            cur.execute("SELECT create_hypertable('data', by_range('time'))")
+            con.commit()
+
+            count = 0
+            for i in range(1, BLOB_COUNT):
+                cur.execute(
+                    "INSERT INTO data (time, blob_data) VALUES (%s, %s);",
+                    (
+                        datetime.now(),
+                        psycopg2.Binary(CHUNK),
+                    ),
+                )
+                count += BLOB_SIZE
+
+    return count
+```
+
+Now we need to create a function to read all the blobs from the table, it is quite easy and we need only one `SELECT` request:
+
+```python
+def read_from_timescale(t1, t2):
+    count = 0
+    with psycopg2.connect(CONNECTION + "/benchmark") as con:
+        with con.cursor() as cur:
+            cur.execute(
+                "SELECT blob_data FROM data WHERE time >= %s AND time < %s;",
+                (datetime.fromtimestamp(t1), datetime.fromtimestamp(t2)),
+            )
+            while True:
+                obj = cur.fetchone()
+                if obj is None:
+                    break
+                count += len(obj[0])
+
+    return count
+```
+
+As you can see to work with Timescale is quite easy if you familiar to any SQL database. However, you may notice that we should put the whole blob into the `INSERT`  request. Of course, it won’t work well for big blobs.
```

Comment:
could you add why it doesn't work well for big blobs and what would be the consequence of that like storing in chunks

Reply:
> As you can see, working with Timescale is quite easy if you are familiar with any SQL database. However, you may notice that we should put the whole blob into the `INSERT`  request. This could cause a performance problem for large blobs because we have to allocate the memory for the request string instead of sending it in chunks.